azure-kusto-data/azure/kusto/data/aio/client.py

```diff
@@ -0,0 +1,95 @@
+from ..exceptions import KustoAioSyntaxError, KustoServiceError
+from ..client import KustoClient as KustoClientSync, _KustoClientBase, KustoConnectionStringBuilder, ClientRequestProperties, ExecuteRequestParams
+from ..response import KustoResponseDataSet
+from ..security import _AadHelper
+
+try:
+    from aiohttp import ClientResponse, ClientSession
+except ImportError:
+    raise KustoAioSyntaxError()
+
+ResponseTuple = namedtuple("ResponseTuple", ["response", "response_json"])
+
+
+@documented_by(KustoClientSync)
+class KustoClient(_KustoClientBase):
+    @documented_by(KustoClientSync.__init__)
+    def __init__(self, kcsb: Union[KustoConnectionStringBuilder, str]):
+        super().__init__(kcsb)
+        # notice that in this context, federated actually just stands for add auth, not aad federated auth (legacy code)
+        self._auth_provider = _AadHelper(self._kcsb, True) if self._kcsb.aad_federated_security else None
+        self._session = ClientSession()
+
+    def __aexit__(self, exc_type, exc_val, exc_tb):
+        self._session.__aexit__(exc_type, exc_val, exc_tb)
+
+    @aio_documented_by(KustoClientSync.execute)
+    async def execute(self, database: str, query: str, properties: ClientRequestProperties = None) -> KustoResponseDataSet:
+        query = query.strip()
+        if query.startswith("."):
+            return await self.execute_mgmt(database, query, properties)
+        return await self.execute_query(database, query, properties)
+
+    @aio_documented_by(KustoClientSync.execute_query)
+    async def execute_query(self, database: str, query: str, properties: ClientRequestProperties = None) -> KustoResponseDataSet:
+        return await self._execute(self._query_endpoint, database, query, None, KustoClient._query_default_timeout, properties)
+
+    @aio_documented_by(KustoClientSync.execute_mgmt)
+    async def execute_mgmt(self, database: str, query: str, properties: ClientRequestProperties = None) -> KustoResponseDataSet:
+        return await self._execute(self._mgmt_endpoint, database, query, None, KustoClient._mgmt_default_timeout, properties)
+
+    @aio_documented_by(KustoClientSync.execute_streaming_ingest)
+    async def execute_streaming_ingest(
+        self,
+        database: str,
+        table: str,
+        stream: io.IOBase,
+        stream_format: Union[DataFormat, str],
+        properties: ClientRequestProperties = None,
+        mapping_name: str = None,
+    ):
+        stream_format = stream_format.value if isinstance(stream_format, DataFormat) else DataFormat(stream_format.lower()).value
+        endpoint = self._streaming_ingest_endpoint + database + "/" + table + "?streamFormat=" + stream_format
+        if mapping_name is not None:
+            endpoint = endpoint + "&mappingName=" + mapping_name
+
+        await self._execute(endpoint, database, None, stream, self._streaming_ingest_default_timeout, properties)
+
+    async def _query(self, endpoint: str, **kwargs) -> list:
```

Comment:
this is confusing, why split _execute to three methods? (_execute, _query, _run_request)